packages/editor-ui/src/mixins/pinData.ts

```diff
@@ -83,22 +93,103 @@ export const pinData = defineComponent({
 				return false;
 			}
 		},
-		isValidPinDataSize(data: string | object): boolean {
+		isValidPinDataSize(data: string | object, activeNodeName: string): boolean {
 			if (typeof data === 'object') data = JSON.stringify(data);
 
-			if (
-				this.workflowsStore.pinDataSize + stringSizeInBytes(data) >
-				MAX_WORKFLOW_PINNED_DATA_SIZE
-			) {
+			const { pinData: currentPinData, ...workflow } = this.workflowsStore.getCurrentWorkflow();
+			const workflowJson = jsonStringify(workflow, { replaceCircularRefs: true });
+
+			const newPinData = { ...currentPinData, [activeNodeName]: data };
+			const newPinDataSize = this.workflowsStore.getPinDataSize(newPinData);
+
+			let isValid = true;
+			if (newPinDataSize > MAX_PINNED_DATA_SIZE) {
 				this.showError(
 					new Error(this.$locale.baseText('ndv.pinData.error.tooLarge.description')),
 					this.$locale.baseText('ndv.pinData.error.tooLarge.title'),
 				);
 
-				return false;
+				isValid = false;
+			}
+
+			if (
+				isValid &&
+				stringSizeInBytes(workflowJson) + newPinDataSize >
+					MAX_WORKFLOW_SIZE - MAX_EXPECTED_REQUEST_SIZE
+			) {
+				this.showError(
+					new Error(this.$locale.baseText('ndv.pinData.error.tooLargeWorkflow.description')),
+					this.$locale.baseText('ndv.pinData.error.tooLargeWorkflow.title'),
+				);
+
+				isValid = false;
+			}
+
+			return isValid;
+		},
+		setPinData(node: INodeUi, data: string | INodeExecutionData[], source: PinDataSource): boolean {
+			if (typeof data === 'string') {
+				if (!this.isValidPinDataJSON(data)) {
+					this.onDataPinningError({ errorType: 'invalid-json', source });
+					throw new Error('Invalid JSON');
+				}
+
+				data = jsonParse(data);
 			}
 
-			return true;
+			if (!this.isValidPinDataSize(data, node?.name ?? '')) {
```

Comment:
This should be defined at this point?
```suggestion
			if (!this.isValidPinDataSize(data, node.name)) {
```

Reply:
Fixed!